In [2]:
from typing import Callable
import numpy, math

# Problem 4

Write a `MATLAB` funcion, called `bisection_method` that takes as inputs a function $f$, two numbers $a$, $b$, an error tolerance, `tol`, and a maximum number of iterations, $N$, and finds a root $c$ of $f$ in the interval $[a , b]$ using the bisection method. Your function should compute a bound on the error, and stop when the error is less than the tolerance, or if the number of iterations exceeds $N$ - whichever happens first.

## Function Definition

In [3]:
# function assumes max iterations is 100 if another maximum is not provided
def bisection_method(f: Callable, a: float, b: float, tol: float, N: int = 100) -> (float, int, float):
    # make sure a, b, and tol are floats
    numbers = [int, float]
    if(type(a) not in numbers or type(b) not in numbers or type(tol) not in numbers):
        print("a, b, and tol must be numbers")
        return

    # make sure N is an integer
    if(type(N) != int or N < 1):
        print("N must be a natural number.")
        return

    # make sure f(x) is callable
    if(not callable(f)):
        print("f(x) must be callable (e.g., a function)")
        return

    # make sure bisection method is cool
    if(f(a) * f(b) > 0):
        return

    # start iterating
    err = (b - a) / 2
    for i in range(N):
        # start evaluating at the halfway point
        c = (a + b) / 2
        
        # if we're in the tolerance range, exit the function
        if(err < tol):
            return(c, i + 1, err)

        # if f(c) == 0, we found a root
        if(f(c) == 0):
            return(c, i + 1, err)

        # otherwise, narrow down the search range
        if(f(a) * f(c) < 0):
            b = c
        else:
            a = c
        err = (b - a) / 2

    # if we exceeded maximum iterations, just return what we 
    # found
    # if I were putting this in some larger code base, I would 
    # probably include whether we exceeded the maximum
    # iterations in the return value, since the function might
    # get used in a way such that the user can't be expected to
    # look at the terminal (e.g., as a module)
    # ---------------------------------------------------------
    print("Maximum iterations exceeded.")
    return(c, i, err)

## Problem 4(c)

In [4]:
f = lambda x: (2*x**3 + 3*x - 1)*math.cos(x) - x
a = -1
b = 1
tol = 1e-5 # python allows E notation for floats

root, iterations, error = bisection_method(f, a, b, tol)

### Problem 4(c.i)

What is the number, $n$, of iterations used?

In [5]:
iterations

18

### Problem 4(c.ii)

What is the error bound?

In [6]:
error

7.62939453125e-06

This is approximately equivalent to $7.6294 \times 10^{-6}$

### Problem 4(c.iii)

The expected rate of convergence for the bisection method can be expressed as $|x_n - c| \leq \frac{b - a}{2^n}$ where $n \in \mathbb{N}$. The big-O notation for this is $O(2^{-n})$

## Problem 4(c.iv)

In [6]:
# plotting; come back to this

# Problem 5

Write a `MATLAB` function, called `fixed_point_method` that takes as inputs a function, $g$, an initial guess $x_0$, an error tolerance, `tol`, and a maximum number of iterations, $N$, and outputs the fixed point of g, obtained using the fixed point method, starting with $x_0$. Your function should have an error defined by $E = |x_n - x_{n-1}|$, which is the absolute difference between the last two iterations, and stop when that error is less than the tolerance, or if the number of iterations exceeds $N$ - whichever happens first. Your function header should look something like:

`function [c, n, err] = fixed_point_method(g, x0, tol, N)`

In [7]:
# function assumes x_0 = 0, tol = 10^-5, and N = 100 if no
# other values are specified
# ---------------------------------------------------------
def fixed_point_method(g: Callable, x_0: float = 0, tol: float = 1e-5, N: int = 100) -> (float, int, float):
    # check if g is a function
    if(not callable(g)):
        print("g(x) must be callable (e.g., a function).")
        return

    # check if x_0 and tol are numbers
    numbers = [int, float]
    if(type(x_0) not in numbers or type(tol) not in numbers):
        print("x_0 and tol must be numbers.")
        return

    # check if N is a natural number
    if(type(N) != int or N < 1):
        print("N must be a natural number")
        return

    # start iterating
    for i in range(N):
        # evaluate at current guess
        c = g(x)
        if(abs(c - x) < tol):
            return(x, 

SyntaxError: incomplete input (1410303702.py, line 26)